### ISSUE TITLE
Finding the training very slow

### ISSUE THREAD

**Issue by ronaldt-git** [PWhiddy/PokemonRedExperiments#74]
I wanted to start from the init state and with no prior training. Ran it over night for millions of steps across 24 cpus and lots of iterations and the ai still only makes it out of the starting room sometimes. Its confusing to me because the reward sum when it leaves the room is often higher than ~.27 and when it doesn't the most it reaches is about .17.

What it accomplished in those 8 hours was figure out how to grab the potion from the PC, which is kinda cool but not even sure why it does it since there's no reward mechanism for adding heal items.

Is there something I can change for the speed of learning from optimal paths? I made plenty of changes to the code but I can't pinpoint the issue. All I can think of is I decreased the model.learn() from multiplying by a factor of 1000 to 10, because it'd didn't save the model at all since it was taking so many hours to go through one learn_step. But there's gotta be some kind of tuning that could help make it learn from rewarding paths more.

```
FILE_PATH = "output_for_simulations.txt"
num_cpu = 24 #64 #46  # Also sets the number of episodes per training iteration
env = None
global model
model = None
global currentLearnStep

# Get all saved models
saved_models = glob.glob(f'session_*/pokemans_*.zip')
saved_models.sort(key=lambda x: int(Path(x).stem.split("_")[1]), reverse=True)


# Initialize the output text file with placeholders
with open(FILE_PATH, 'w') as file:
    for rank in range(num_cpu):
        file.write(f"Rank {rank} - initializing...\n")

lock = multiprocessing.Lock()  # You're using multiprocessing, so use its Lock


def signal_handler(signum, frame):
    print("Interrupted! Closing environments...")
    save_model()
    if env is not None:
        env.close()
    sys.exit(0)

# Set the signal handler
signal.signal(signal.SIGINT, signal_handler)


def make_env(rank, env_conf, should_render = False, seed=0, pretraining_steps=0):
    def _init():
        env = RedGymEnv(env_conf, should_render, rank=rank, lock=lock, file_path=FILE_PATH, original_steps=pretraining_steps, ep_length=ep_length, num_ranks=num_cpu)  # Passing lock and file path
        env.reset(seed=(seed + rank))
        return env
    set_random_seed(seed)
    return _init


def save_model():
    if model is not None:
        print("current learn step +1: ", currentLearnStep + 1)
        print("pretraining: ", pretraining_steps)
        total_steps_so_far = pretraining_steps + ep_length*num_cpu*(currentLearnStep+1)
        print("total steps so far: ", total_steps_so_far)
        model_path = os.path.join(sess_path, f"pokemans_{total_steps_so_far}_steps")
        model.save(model_path)
        print("Saving Most recent model at: " ,total_steps_so_far, " steps.")


if __name__ == '__main__':

    ep_length = 512
    if saved_models:
        pretraining_steps = int(re.search('pokemans_(\d+)_steps', saved_models[0]).group(1))
    else:
        pretraining_steps = 0

    sess_path = f'session_{str(uuid.uuid4())[:8]}'
    if not os.path.exists(sess_path):
        os.makedirs(sess_path)
    args = get_args('run_baseline_parallel.py', ep_length=ep_length, sess_path=sess_path)

    env_config = {
                'headless': False, 'save_final_state': True, 'early_stop': False,
                'action_freq': 24, 'init_state': '../init.state', 'max_steps': ep_length, 
                'print_rewards': True, 'save_video': False, 'fast_video': True, 'session_path': sess_path,
                'gb_path': '../PokemonRed.gb', 'debug': False, 'sim_frame_dist': 2_000_000.0
            }
    
    env_config = change_env(env_config, args)
    
    env = SubprocVecEnv([make_env(i, env_config, should_render=(i == 0), pretraining_steps=pretraining_steps) for i in range(num_cpu)])
    
    #checkpoint_callback = CheckpointCallback(save_freq=ep_length, save_path=sess_path,
    #                                 name_prefix='poke')
    #env_checker.check_env(env)
    learn_steps = 20

    print("checking if saved models exist: ", saved_models)
    if saved_models:
        
        print(f'\nloading model: {saved_models[0]}')
        model = PPO.load(saved_models[0], env=env)
        print("Starting at Model with ",pretraining_steps, " steps.")
        model.n_steps = ep_length
        model.n_envs = num_cpu
        model.rollout_buffer.buffer_size = ep_length
        model.rollout_buffer.n_envs = num_cpu
        model.rollout_buffer.reset()
    else:
        print('\n Loading untrained Model: ')
        model = PPO('CnnPolicy', env, verbose=1, n_steps=ep_length, batch_size=512, n_epochs=1, gamma=0.999)
        pretraining_steps = 0
    
    for currentLearnStep in range(learn_steps):
        print("entering currentLearnStep: ", currentLearnStep)
        model.learn(total_timesteps=(ep_length)*num_cpu*10) #, callback=checkpoint_callback

        # Save the model
        save_model()
        
        
        # Render steps after learning
        obs = env.reset()
        for _ in range(100):  # Render 100 steps, adjust as needed
            env.render()
            action, _ = model.predict(obs)
            obs, _, _, _ = env.step(action)
        time.sleep(1)  # Pause for a second between iterations, adjust as needed

```

**Comment by setomage** [comment]
I was actually wounder how long a training season should take. Been letting it run for almost 24 hours, and I can't find anything that says that the system auto saves the season if I just end it, or if I should let it run till it's done.

**Comment by techmore** [comment]
@setomage you can edit your "run_baseline_parallel.py"  with this patch #76 and on the next run you you will see "loading checkpoint" as it resumes from the folder with the newest timestamp.

**Comment by setomage** [comment]
> @setomage you can edit your "run_baseline_parallel.py" with this patch #76 and on the next run you you will see "loading checkpoint" as it resumes from the folder with the newest timestamp.

@techmore So does this mean I can just end the current training session, and not lose anything? I ask since the folder for the current training session is also at 24GB

**Comment by techmore** [comment]
Thats a fair point and while I can't guarantee anything I can say I'm been using this for a few days on my laptop and I can show you a picture of me using break and resuming and it being the same session.

I actually seem to be able to keep my session running, edit the file and run it in a new session as well without suspending the first session. My machine just grinds to a halt. 

The is the effective result of that patch.
<img width="820" alt="Screenshot 2023-10-18 at 8 40 25 PM" src="https://github.com/PWhiddy/PokemonRedExperiments/assets/9263712/b2a977b4-ba85-4d30-9df1-479868088a94">

<img width="784" alt="Screenshot 2023-10-18 at 8 47 17 PM" src="https://github.com/PWhiddy/PokemonRedExperiments/assets/9263712/b36c2e7b-7247-4e84-81d8-1920833a785c">

**Comment by setomage** [comment]
@techmore Tried the changes that you linked too, sadly as a filthy Windows user, it doesn't work. 

File "\PokemonRedExperiments-master\baselines\run_baseline_parallel2.py", line 69, in <module>
    session_folder, latest_poke_file = find_latest_session_and_poke()
  File "\PokemonRedExperiments-master\baselines\run_baseline_parallel2.py", line 30, in find_latest_session_and_poke
    all_folders = baselines()
NameError: name 'baselines' is not defined

**Comment by techmore** [comment]
@setomage Thank you for the update and sorry to hear that. I looks to be the issue could be "all_folder = baselines()" which I think may be a mistake. While I'm trying to find that error I'm posted my local code in a link below that you may want to try in the meantime.

This code includes tensorboard  and changes  num_cpu = os.cpu_count() instead of 44.

[https://raw.githubusercontent.com/techmore/PokemonRedExperiments/patch-1/baselines/run_baseline_parallel.py](https://raw.githubusercontent.com/techmore/PokemonRedExperiments/patch-1/baselines/run_baseline_parallel.py)

sorry about that... lets try to get this working for everyone

**Comment by setomage** [comment]
> @setomage Thank you for the update and sorry to hear that. I looks to be the issue could be "all_folder = baselines()" which I think may be a mistake. While I'm trying to find that error I'm posted my local code in a link below that you may want to try in the meantime.
> 
> This code includes tensorboard and changes num_cpu = os.cpu_count() instead of 44.
> 
> https://raw.githubusercontent.com/techmore/PokemonRedExperiments/patch-1/baselines/run_baseline_parallel.py
> 
> sorry about that... lets try to get this working for everyone

Swing and a miss for Windows users. "os is not defined"   LOL

I've been slowly working on the code for Windows and getting to run better. Before I was running at a 100% on all 24 cores and now I'm at about 80 - 85%. Off the hope I would also be running 35GB of RAM, and got it down to 26GB. 

Still had to hard point to my last training data and its zip, just hoping that since my last training session had four of them, that the latest one took everything from the last three...  ^_^" 

I know from looking at the code that 'run_baseline' and the 'run_pretrained_interactive' is also meant to look for the training data and be hard coded into the files. So if they're not edited every run time, you only get the old builds. 

Once I get things running smoothly on my end, I am planing on doing a Windows build and hoping I can do what you're doing and making it so the code looks for the most resent files on its own and not have to be hard coded.

*Yoda Voice*
Out of box and noob friendly, this is not.

**Comment by RussellMaggs** [comment]
You need to an import for os at the top of the file like so

```
from os.path import exists
from pathlib import Path
import uuid
from red_gym_env import RedGymEnv
from stable_baselines3 import A2C, PPO
from stable_baselines3.common import env_checker
from stable_baselines3.common.vec_env import DummyVecEnv, SubprocVecEnv
from stable_baselines3.common.utils import set_random_seed
from stable_baselines3.common.callbacks import CheckpointCallback
from argparse_pokemon import *
import os
```

Hopefully that should fix the os is not defined error

**Comment by flowflows** [comment]
I tried training from scratch and found the same issue, I think the reward in pixel change could be why it stays within the first map early in training. Because I've found it can move without issue to something until it causes dialog interactions and then gets fixated on those. So when it eventually finds a way to leave the area I think it starts understanding there are ways to find more areas but takes wayyyy too long to get there, and it is likely how it found every single zone thus far, which could explain some of it's navigation choices.
